sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala

```diff
@@ -337,7 +337,7 @@ case class StateStoreRestoreExec(
     if (keyExpressions.isEmpty) {
       AllTuples :: Nil
     } else {
-      ClusteredDistribution(keyExpressions, stateInfo.map(_.numPartitions)) :: Nil
+      StatefulOpClusteredDistribution(keyExpressions, stateInfo.map(_.numPartitions)) :: Nil
```

Comment:
I am wondering if this change could introduce extra shuffle for streaming aggregate. Previously the operator requires `ClusteredDistribution`, and right now it requires `StatefulOpClusteredDistribution`/`HashClusteredDistribution`. 

`ClusteredDistribution` is more relaxed than `HashClusteredDistribution` in the sense that a `HashPartitioning(c1)` can satisfy `ClusteredDistribution(c1, c2)`, but cannot satisfy `HashClusteredDistribution(c1, c2)`. In short, `ClusteredDistribution` allows child to be hash-partitioned on subset of required keys. So for aggregate, if the plan is already shuffled on subset of group-by columns, Spark will not add a shuffle again before group-by.

For example:

```
MemoryStream[(Int, Int)].toDF()
  .repartition($"_1")
  .groupBy($"_1", $"_2")
  .agg(count("*"))
  .as[(Int, Int, Long)]
```

and the query plan:

```
WriteToDataSourceV2 org.apache.spark.sql.execution.streaming.sources.MicroBatchWrite@5940f7c2, org.apache.spark.sql.execution.datasources.v2.DataSourceV2Strategy$$Lambda$1940/1200613952@4861dac3
+- *(4) HashAggregate(keys=[_1#588, _2#589], functions=[count(1)], output=[_1#588, _2#589, count(1)#596L])
   +- StateStoreSave [_1#588, _2#589], state info [ checkpoint = file:/private/var/folders/y5/hnsw8mz93vs57ngcd30y6y9c0000gn/T/streaming.metadata-0d7cb004-92dd-4b0d-9d90-5a65c0d2934c/state, runId = 68598bd1-cf35-4bf7-a167-5f73dc9f4d84, opId = 0, ver = 0, numPartitions = 5], Complete, 0, 1
      +- *(3) HashAggregate(keys=[_1#588, _2#589], functions=[merge_count(1)], output=[_1#588, _2#589, count#663L])
         +- StateStoreRestore [_1#588, _2#589], state info [ checkpoint = file:/private/var/folders/y5/hnsw8mz93vs57ngcd30y6y9c0000gn/T/streaming.metadata-0d7cb004-92dd-4b0d-9d90-5a65c0d2934c/state, runId = 68598bd1-cf35-4bf7-a167-5f73dc9f4d84, opId = 0, ver = 0, numPartitions = 5], 1
            +- *(2) HashAggregate(keys=[_1#588, _2#589], functions=[merge_count(1)], output=[_1#588, _2#589, count#663L])
               +- *(2) HashAggregate(keys=[_1#588, _2#589], functions=[partial_count(1)], output=[_1#588, _2#589, count#663L])
                  +- Exchange hashpartitioning(_1#588, 5), REPARTITION_BY_COL, [id=#2008]
                     +- *(1) Project [_1#588, _2#589]
                        +- MicroBatchScan[_1#588, _2#589] MemoryStreamDataSource
```

One can argue the previous behavior for streaming aggregate is not wrong. As long as all rows for same keys are colocated in same partition, `StateStoreRestore/Store` should output correct answer for streaming aggregate. If we make the change here, I assume one extra shuffle on `($"_1", $"_2")` would be introduced, and it might yield incorrect result when running the new query plan against the existing state store?

Reply:
But we don't define "adding repartition" as "unsupported operation across query lifetime", no?

The thing is that once the query is run once, the partitioning of stateful operator must not be changed during lifetime. Since we don't store the information of partitioning against stateful operator in the checkpoint, we have no way around other than enforcing the partitioning of stateful operator as the "one" what we basically expect.

As I said in #32875, there is a room for improvement, but the effort on improvement must be performed after we fix this issue.

Reply:
> But we don't define "repartition just before stateful operator" as "unsupported operation across query lifetime", no?


I ran the query in `StreamingAggregationSuite.scala` and it seems fine. I briefly checked `UnsupportedOperationChecker.scala` and didn't find we disallow "repartition just before stateful operator".

> The thing is that once the query is run, the partitioning of stateful operator must not be changed during lifetime. Since we don't store the information of partitioning against stateful operator in the checkpoint, we have no way around other than enforcing the partitioning of stateful operator as the "one" what we basically expect.


I agree with you @HeartSaVioR. I want to raise a concern here it might change the query plan for certain streaming aggregate query (as above synthetic query), and it could break existing state store when running with next Spark 3.3 code, based on my limited understanding.

Reply:
> If we make the change here, I assume one extra shuffle on ($"_1", $"_2") would be introduced, and it might yield incorrect result when running the new query plan against the existing state store?

Unfortunately yes. We may need to craft some tools to analyze the state and repartition if the partitioning is already messed up. But leaving this as it is would bring more chances to let users' state be indeterministic.

Reply:
>> But we don't define "repartition just before stateful operator" as "unsupported operation across query lifetime", no?

> I ran the query in StreamingAggregationSuite.scala and it seems fine. I briefly checked UnsupportedOperationChecker.scala and didn't find we disallow "repartition just before stateful operator".

That is the problem we have. We didn't disallow the case where it brings silent correctness issue.

Reply:
> That is the problem we have. We didn't disallow the case where it brings silent correctness issue.

This is just a synthetic example I composed to verify my theory. But I think it might break for more cases, such as `GROUP BY c1, c2` after `JOIN ON c1`. I am trying to say it would break for the queries which are partitioned on subset of group-by keys.

Reply:
Even nowadays multiple stateful operators don't work properly due to the global watermark, so we don't need to worry about the partitioning between stateful operators. We just need to worry about the partitioning between upstream (non stateful) and the stateful operator.